## Real Log data

In [2]:
import pandas as pd

logon = pd.read_csv("/Users/dholakiyan/Desktop/insider-threat-detection/data/raw/r4.2/logon.csv")
logon['date'] = pd.to_datetime(logon['date'])

print(logon.shape)
print(logon.head(10))
print(logon['activity'].value_counts())

(854859, 5)
                         id                date     user       pc activity
0  {X1D9-S0ES98JV-5357PWMI} 2010-01-02 06:49:00  NGF0157  PC-6056    Logon
1  {G2B3-L6EJ61GT-2222RKSO} 2010-01-02 06:50:00  LRR0148  PC-4275    Logon
2  {U6Q3-U0WE70UA-3770UREL} 2010-01-02 06:53:04  LRR0148  PC-4124    Logon
3  {I0N5-R7NA26TG-6263KNGM} 2010-01-02 07:00:00  IRM0931  PC-7188    Logon
4  {D1S0-N6FH62BT-5398KANK} 2010-01-02 07:00:00  MOH0273  PC-6699    Logon
5  {S6P1-M4MK04BB-0722IITW} 2010-01-02 07:07:00  LAP0338  PC-5758    Logon
6  {M6O6-F9UU11TJ-2166YVFU} 2010-01-02 07:08:00  MHH0180  PC-9822    Logon
7  {D0P1-U5IU49QG-5472TMLI} 2010-01-02 07:08:00  NOB0181  PC-3446    Logon
8  {F9K3-Q1PD92IU-5821IGXW} 2010-01-02 07:13:00  AHC0142  PC-8893    Logon
9  {S9A0-A3DP89HG-6719JORU} 2010-01-02 07:14:00  CTR0341  PC-6184    Logon
activity
Logon     470591
Logoff    384268
Name: count, dtype: int64


## Checking activity pattern of Suspicious User

In [3]:
answers = pd.read_csv("/Users/dholakiyan/Desktop/insider-threat-detection/data/processed/answers_r42.csv")
sample_user = answers['user'].iloc[0]
print("Checking user:", sample_user)

suspect_logons = logon[logon['user'] == sample_user].sort_values('date')
print(suspect_logons)

Checking user: AAM0658
                              id                date     user       pc  \
1403    {D6U0-S8PP45WQ-2670HHAA} 2010-01-04 09:14:00  AAM0658  PC-9923   
2778    {L7T0-O8SG89IO-8831RXJZ} 2010-01-04 21:19:00  AAM0658  PC-9923   
3973    {O5E8-I8TD55VX-1793UOBV} 2010-01-05 09:21:00  AAM0658  PC-9923   
5388    {P8Y4-S3NR31ZM-2748TRCK} 2010-01-05 21:15:00  AAM0658  PC-9923   
6521    {O7M3-U7HP52RB-8417ZOTI} 2010-01-06 09:05:00  AAM0658  PC-9923   
...                          ...                 ...      ...      ...   
571387  {B7F3-V3KV45ET-2293FRFQ} 2010-11-15 21:11:00  AAM0658  PC-9923   
572489  {M6N3-W8WU73BZ-9884QBLI} 2010-11-16 09:13:00  AAM0658  PC-9923   
573753  {B0S3-I6KL21OV-2947HATX} 2010-11-16 21:19:00  AAM0658  PC-9923   
574764  {S3K3-X2AO27JZ-5561IIXB} 2010-11-17 09:00:00  AAM0658  PC-9923   
576076  {P6W8-O6LP29IQ-0152GNAZ} 2010-11-17 21:15:00  AAM0658  PC-9923   

       activity  
1403      Logon  
2778     Logoff  
3973      Logon  
5388     Logoff 

## Checking activity of normal user:

In [4]:
normal_user = logon['user'].sample(1, random_state=1).values[0]
normal_logons = logon[logon['user'] == normal_user].sort_values('date')
print(normal_logons.head(10))

                            id                date     user       pc activity
928   {G8Z4-Y6TO08HP-8436XRVV} 2010-01-04 07:58:00  LDG0459  PC-5621    Logon
2158  {U0D5-N1KA98LD-7422ZJJA} 2010-01-04 17:14:48  LDG0459  PC-4124    Logon
2198  {Q2I2-B3AW50NS-2611WRAQ} 2010-01-04 17:20:04  LDG0459  PC-4124   Logoff
2365  {E7W4-C1DE62UA-7487OIKA} 2010-01-04 18:05:00  LDG0459  PC-5621   Logoff
3428  {N7P9-I9PF56LK-7046BIBL} 2010-01-05 08:02:00  LDG0459  PC-5621    Logon
4218  {P0S3-B6PX78SC-9371RHPD} 2010-01-05 13:28:36  LDG0459  PC-5621    Logon
5001  {V9D9-G5QC53QU-7451RMJC} 2010-01-05 18:00:00  LDG0459  PC-5621   Logoff
6018  {J9H3-K4EQ65FT-6271VKSI} 2010-01-06 07:57:00  LDG0459  PC-5621    Logon
6729  {R5D4-E3LG95OU-2886PMLI} 2010-01-06 12:31:02  LDG0459  PC-5621    Logon
7442  {J8J5-I4JB11CE-7979JEIE} 2010-01-06 17:43:59  LDG0459  PC-4124    Logon


Malicious user (AAM0658) — Scenario 1 suspect
Look at the pattern 
Always logs on ~9:00-9:20 AM, logs off ~21:15-21:19 PM — every single day, like clockwork
Only ever uses ONE machine: PC-9923
Exactly 2 events per day (one Logon, one Logoff) — extremely regular, almost robotic

Normal user (LDG0459) — random pick (not suspicious user)
Logs on/off multiple times per day (not just once) — e.g., Jan 4th: on at 07:58, on again at 17:14, off at 17:20, off at 18:05
Uses TWO different machines: PC-5621 and PC-4124
More natural, human variability in timing (07:58, 08:02, 07:57 — not identical every day)

## Is there a shift in AAM0658's pattern?

In [8]:
suspect_logons['hour'] = suspect_logons['date'].dt.hour
print(suspect_logons.groupby(suspect_logons['date'].dt.to_period('M'))['hour'].agg(['mean','min','max']))

              mean  min  max
date                        
2010-01  15.000000    9   21
2010-02  15.000000    9   21
2010-03  15.000000    9   21
2010-04  15.000000    9   21
2010-05  15.000000    9   21
2010-06  15.000000    9   21
2010-07  15.000000    9   21
2010-08  15.000000    9   21
2010-09  15.000000    9   21
2010-10  13.040000    0   21
2010-11  14.107143    2   21


### Observation: 

January through September (2010-01 to 2010-09): rock-solid consistent

Mean hour: exactly 15.0 every single month
Min: always 9 (9 AM — earliest logon)
Max: always 21 (9 PM — latest logoff)

This user had an extremely stable, predictable routine for 9 straight months. Textbook normal behavior.
October (2010-10): something changes

Mean drops to 13.04 (pulled earlier/more spread out)
Min drops to 0 — meaning at least one logon happened at midnight (hour 0)
Max stays at 21

November (2010-11): continues the shift

Mean at 14.1
Min at 2 — activity as early as 2 AM
Max still 21

What this means
Starting in October, this user began logging on/off at hours (midnight, 2 AM) that never once happened in the previous 9 months. This is precisely the "after-hours activity that didn't exist before" signal from Scenario 1's description — and you found it just by grouping timestamps by month. So behavioral deviation from an established personal baseline is detectable, even with something as simple as monthly hour statistics. This is exactly what autoencoder will later learn to do automatically, at scale, across thousands of users and dozens of features — but signal exists.

## After hour activity :

In [9]:
after_hours = suspect_logons[(suspect_logons['hour'] < 6)]
print(after_hours)

                              id                date     user       pc  \
527805  {Y6R0-G0UT09SU-3761DVZN} 2010-10-21 00:18:18  AAM0658  PC-9923   
527818  {I2K5-C9ZW78EP-6081NVTR} 2010-10-21 01:20:00  AAM0658  PC-9923   
532577  {K3V4-Y4OK65SI-1583GEOQ} 2010-10-23 01:34:19  AAM0658  PC-9923   
537623  {Q7J2-R0UY09AF-7403UNWV} 2010-10-27 04:31:52  AAM0658  PC-9923   
537647  {Y8T3-R7YX37UT-0796DLDV} 2010-10-27 05:54:39  AAM0658  PC-9923   
542318  {S8Q1-Z1OG01ED-8756DWPR} 2010-10-29 00:06:47  AAM0658  PC-9923   
542414  {J6L0-W5OJ26KB-8630RYVJ} 2010-10-29 05:23:28  AAM0658  PC-9923   
547366  {B4H0-T0VS68NI-2874IXYP} 2010-11-02 02:31:32  AAM0658  PC-9923   
547372  {Z1O0-J1QH34XQ-4354CGWO} 2010-11-02 03:02:53  AAM0658  PC-9923   

       activity  hour  
527805    Logon     0  
527818   Logoff     1  
532577    Logon     1  
537623    Logon     4  
537647   Logoff     5  
542318    Logon     0  
542414   Logoff     5  
547366    Logon     2  
547372   Logoff     3  


### Observation:
Notice these aren't randomly scattered — they cluster tightly in a ~2 week window (Oct 21 – Nov 2), all on their usual single machine (PC-9923), all overnight. This lines up exactly with the answer key's start timestamp for this user 

## worth double-checking against answers_r42

In [10]:
print(answers[answers['user'] == 'AAM0658'][['start', 'end']])

                 start                  end
0  2010-10-23 01:34:19  2010-10-29 05:23:28


### Observation:

"hour < 6" filter caught the exact official start and end timestamps perfectly (that's not a coincidence — it confirms detection logic is genuinely sound). But it also caught a couple of extra sessions outside the officially labeled window (Oct 21, and Nov 2).